dvc/repo/open_repo.py

```diff
@@ -97,9 +97,22 @@ def clean_repos():
         _remove(path)
 
 
-def _get_remote_config(url):
+def _get_remote_config(url, *args, **kwargs):
     try:
-        repo = Repo(url)
+        # Get the config dict passed from the parent call, default to empty dict
+        user_config = kwargs.get("config", {})
+        # It seems some tests might be passing a 'config' key that is not a dict
+        if not isinstance(user_config, dict):
+            user_config = {}
+        no_scm_flag = user_config.get("core", {}).get("no_scm", None)
+
+        if no_scm_flag is not None:
+            # Honour specific SCM treatment if requested in the call
+            repo = Repo(url, config={"core": {"no_scm": no_scm_flag}})
```

Comment:
AFAIR, `Repo(config=...)` should just work.
```suggestion
            repo = Repo(url, config=kwargs.get("config"))
```

I don't want to specialize `core.no_scm` in any way or handle it.

Reply:
I agree that it doesn't feel ideal to handle `core.no_scm` itself.  Your solution makes sense and it works for my specific use case, but it triggers other errors in the `dvc` test suite; which makes me think there are other non-`core.no_scm` configuration options that are being used that `_get_remote_config()` doesn't like.

(I went with the `core.no_scm` specific approach to highlight the need.)

These are the errors I get with using `repo = Repo(url, config=kwargs.get("config"))`:
```console
FAILED tests/func/test_import.py::test_import_no_hash[files1-expected_info_calls1] - dvc_data.index.index.DataIndexDirError: failed to load directory ('dir',)
FAILED tests/func/test_repo_index.py::test_data_index - dvc_data.index.index.DataIndexDirError: failed to load directory ('edir',)
FAILED tests/func/repro/test_repro_pull.py::test_repro_pulls_missing_import - dvc.exceptions.ReproductionError: failed to reproduce 'foo.dvc'
FAILED tests/func/test_data_cloud.py::test_pull_external_dvc_imports - FileNotFoundError: [Errno 2] No storage files available: 'foo'
FAILED tests/func/test_data_cloud.py::test_pull_external_dvc_imports_mixed - FileNotFoundError: [Errno 2] No storage files available: 'foo'
FAILED tests/func/repro/test_repro.py::test_repro_pulls_missing_import - dvc.exceptions.ReproductionError: failed to reproduce 'foo.dvc'
FAILED tests/func/test_import.py::test_import_dir - dvc_data.index.index.DataIndexDirError: failed to load directory ('dir',)
FAILED tests/func/test_import.py::test_import_file_from_dir - dvc_data.index.index.DataIndexDirError: failed to load directory ('dir',)
FAILED tests/func/test_import.py::test_import_file_from_dir_to_dir - dvc_data.index.index.DataIndexDirError: failed to load directory ('dir',)
FAILED tests/func/test_import.py::test_import_rev - FileNotFoundError: [Errno 2] No storage files available: 'foo'
FAILED tests/func/test_import.py::test_pull_imported_stage - dvc.exceptions.CheckoutError: Checkout failed for following targets:
FAILED tests/func/test_import.py::test_pull_import_no_download - FileNotFoundError: [Errno 2] No such file or directory: '/private/var/folders/_3/h8jc4f6d5gg6dwk7t6464_z80000gp/T/pytest-of-rodrigo.goya/pytest-17/popen-gw13/test_pull_import_no_download0/.dvc/cache/fs/local/e3501e821bcee8f40107794afbe767d1/.F0VDC8H_fGFnvnEcrt...
FAILED tests/func/test_import.py::test_pull_import_no_download_rev_lock - dvc.exceptions.DownloadError: 1 files failed to download
FAILED tests/func/test_import.py::test_pull_imported_directory_stage[dir] - dvc_data.index.index.DataIndexDirError: failed to load directory ('dir',)
FAILED tests/func/test_import.py::test_pull_imported_directory_stage[dir/] - dvc_data.index.index.DataIndexDirError: failed to load directory ('dir',)
FAILED tests/func/test_import.py::test_pull_wildcard_imported_directory_stage - dvc_data.index.index.DataIndexDirError: failed to load directory ('dir123',)
FAILED tests/func/test_update.py::test_update_import[True] - FileNotFoundError: [Errno 2] No storage files available: 'version'
FAILED tests/func/test_import.py::test_pull_non_workspace - FileNotFoundError: [Errno 2] No storage files available: 'foo'
FAILED tests/func/test_update.py::test_update_import_after_remote_updates_to_dvc - FileNotFoundError: [Errno 2] No storage files available: 'version'
FAILED tests/func/test_import.py::test_import_with_jobs - dvc_data.index.index.DataIndexDirError: failed to load directory ('dir1',)
```

Reply:
For example, looking closer at one of the failing tests, `tests/func/test_repo_index.py::test_data_index`


It seems that there is a competition between `cache` folders. (Maybe an artifact on how tests are designed?)

- A call to `dvc.repo.imp.py:imp()` eventually triggers a call to `_get_remote_config()` (call stack at the bottom)
- A `url` is provided, and a `**kwargs` that contains a `config` key, which provides information on the cache, including a cache directory.
   - `url` = `/private/var/folders/_3/h8jc4f6d5gg6dwk7t6464_z80000gp/T/pytest-of-rgoya/pytest-27/erepo0`
   - `config` = `{'cache': {'protected': False, 'slow_link_warning': True, 'verify': False, 'dir': '/private/var/folders/_3/h8jc4f6d5gg6dwk7t6464_z80000gp/T/pytest-of-rgoya/pytest-27/test_data_index0/.dvc/cache'}}`

- Besides the remote's name, `_get_remote_config()` returns `repo.cache.local_cache_dir`.

- The presence of `config = {"cache": {"dir": "/path/to/cache"}}` changes the return value of `repo.cache.local_cache_dir`:
   - with `Repo()` alone it is: `/Users/rgoya/code/dvc_fork`
   - with `Repo(config=kwargs.get("config"))` it is: `'/private/var/folders/_3/h8jc4f6d5gg6dwk7t6464_z80000gp/T/pytest-of-rgoya/pytest-27/test_data_index0/.dvc/cache`

![image](https://github.com/user-attachments/assets/0f3341d1-8428-4a77-84c7-aa77e8a3fe3e)

Reply:
> AFAIR, `Repo(config=...)` should just work.
>      - `repo = Repo(url, config={"core": {"no_scm": no_scm_flag}})`
>      + `repo = Repo(url, config=kwargs.get("config"))`
> I don't want to specialize `core.no_scm` in any way or handle it.

@skshetry, Ok, I finally got some time to track down the errors.

I took your suggestion of using something closer to `Repo(url, config=kwargs.get("config"))`.

I tracked the errors that were triggered by that call to the way that `dvc` manages imports. Mainly, it creates a `DVCFileSystem` for a remote repo using a local cache (see issue https://github.com/iterative/dvc/issues/9385 / fix https://github.com/iterative/dvc/pull/9423 ). This means that when `_get_remote_config()` is eventually called, it receives the `url` of the remote repo, but a `kwargs["config"]["cache"]` that points to the local cache. 

Since the function of `_get_remote_config()` is to get the actual remote's cache location, in order for `dvc import` to work, `_get_remote_config()` needs to ignore anything sent in `kwargs["config"]["cache"]`.

I modified the code to to the above, and added some comments to the code for future reference.

I also added a related test to the test suite.

Please let me know what you think.

-----

Note: FYI, I've been seeing some flakiness in the test suite from what seems to be a race condition in the logs, where a message of `DEBUG    s3fs:utils.py:80 RC: discarding all clients` sneaks in and fails the assert, e.g.:
```
>       assert first(caplog.messages) == expected_message
E       assert 'RC: discarding all clients' == "Output 'path... 'stage.dvc'."
E         
E         + RC: discarding all clients
E         - Output 'path'(stage: 'stage.dvc') is missing version info. Cache for it will not be collected. Use `dvc repro` to get your pipeline up to date.
E         - You can also use `dvc commit stage.dvc` to associate existing 'path' with stage: 'stage.dvc'.

tests/unit/output/test_output.py:106: AssertionError
----------------------------- Captured stderr call -----------------------------
WARNING: Output 'path'(stage: 'stage.dvc') is missing version info. Cache for it will not be collected. Use `dvc repro` to get your pipeline up to date.
You can also use `dvc commit stage.dvc` to associate existing 'path' with stage: 'stage.dvc'.
------------------------------ Captured log call -------------------------------
DEBUG    s3fs:utils.py:80 RC: discarding all clients
WARNING  dvc.output:output.py:1160 Output 'path'(stage: 'stage.dvc') is missing version info. Cache for it will not be collected. Use `dvc repro` to get your pipeline up to date.
You can also use `dvc commit stage.dvc` to associate existing 'path' with stage: 'stage.dvc'.
```